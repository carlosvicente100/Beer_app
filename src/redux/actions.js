export const fetchRequested = "FETCH_REQUESTED";
export const fetchRequestedItems = "FETCH_REQUESTED_ITEMS";
export const fetchSuccessItems = "FETCH_SUCCESS_ITEMS";
export const fetchErrorItems = "FETCH_ERROR_ITEMS";

//filters
export const fetchSetDates = "FETCH_SET_DATES";
export const fetchResetDates = "FETCH_RESET_DATES";
