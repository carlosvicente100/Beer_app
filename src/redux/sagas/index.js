import { all } from 'redux-saga/effects'
import fetchDataFromUrl from './fetch'

export default function* rootSaga() {
  yield all([fetchDataFromUrl()])
}
