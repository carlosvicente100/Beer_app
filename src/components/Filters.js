import React, { useState, useEffect } from 'react'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { useDispatch, useSelector } from 'react-redux'
import Select from './Select'
import { fetchSetDates, fetchResetDates } from '../redux/actions'

import { formatOrderYears, formatDate, DEFAULT_VALUE, MonthList } from '../utils'

const FiltersContainer = styled.div`
  background-color: #227093;
  color: white;
  text-align: center;
`
const Label = styled.label`
  width: 50%;
  display: inline-grid;
  grid-area: input;
  grid-template-areas:
    'span'
    'select';
  span {
    grid-area: span;
  }
  div.Select {
    width: 100%;
    grid-area: select;
  }
`
const FilterButton = styled.button`
  background-color: #2980b9;
  border: none;
  color: white;
  padding: 12px 16px;
  font-size: 16px;
  cursor: pointer;
  &:hover {
    background-color: tomato;
  }
  &:disabled {
    background-color: #7f8c8d;
  }
`

const Filters = ({ updateFilters }) => {
  const dispatch = useDispatch()
  const { items } = useSelector((state) => state.itemsReducer)
  const yearList = formatOrderYears(items.map((a) => formatDate(a.first_brewed)))
  const { startDate, endDate } = useSelector((state) => state.filtersReducer)
  const [monthInitial, setMonthInitial] = useState(startDate.month)
  const [yearInitial, setYearhInitial] = useState(startDate.year)
  const [monthFinal, setMonthFinal] = useState(endDate.month)
  const [yearFinal, setYearFinal] = useState(endDate.year)

  const FiltersSetted = () =>
    monthInitial !== DEFAULT_VALUE &&
    yearInitial !== DEFAULT_VALUE &&
    monthFinal !== DEFAULT_VALUE &&
    yearFinal !== DEFAULT_VALUE

  const resetValues = () => {
    updateFilters()
    setMonthInitial(DEFAULT_VALUE)
    setYearhInitial(DEFAULT_VALUE)
    setMonthFinal(DEFAULT_VALUE)
    setYearFinal(DEFAULT_VALUE)
    dispatch({
      type: fetchResetDates
    })
  }
  const setFilter = () => {
    updateFilters()
    dispatch({
      type: fetchSetDates,
      data: {
        startDate: {
          year: parseInt(yearInitial),
          month: parseInt(monthInitial)
        },
        endDate: {
          year: parseInt(yearFinal),
          month: parseInt(monthFinal)
        }
      }
    })
  }
  return (
    <FiltersContainer className="Filters">
      <Label>
        Start Month
        <Select values={MonthList} current={monthInitial} onSelect={setMonthInitial} />
      </Label>
      <Label>
        End Month
        <Select values={MonthList} current={monthFinal} onSelect={setMonthFinal} />
      </Label>

      <Label>
        Start year
        <Select values={yearList} current={yearInitial} onSelect={setYearhInitial} />
      </Label>
      <Label>
        End year
        <Select values={yearList} current={yearFinal} onSelect={setYearFinal} />
      </Label>
      <FilterButton disabled={!FiltersSetted()} onClick={setFilter}>
        Filter
      </FilterButton>
      <FilterButton disabled={!FiltersSetted()} onClick={resetValues}>
        Reset All
      </FilterButton>
    </FiltersContainer>
  )
}

Filters.propTypes = {
  updateFilters: PropTypes.func.isRequired
}

Filters.defaultProps = {}

export default Filters
