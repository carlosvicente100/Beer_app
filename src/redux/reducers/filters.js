import { fetchSetDates, fetchResetDates } from "../actions";

export const initialState = {
  startDate: {
    year: -1,
    month: -1
  },
  endDate: {
    year: -1,
    month: -1
  }
};

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case fetchSetDates:
      return {
        ...state,
        startDate: action.data.startDate,
        endDate: action.data.endDate
      };
    case fetchResetDates:
      return {
        ...initialState
      };
    default:
      return {
        ...state
      };
  }
}
