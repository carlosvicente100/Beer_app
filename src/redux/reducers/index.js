import { combineReducers } from 'redux'

import itemsReducer from './items'
import filtersReducer from './filters'

export default combineReducers({
  itemsReducer,
  filtersReducer
})
