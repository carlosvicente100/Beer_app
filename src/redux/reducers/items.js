import { fetchSuccessItems, fetchErrorItems } from '../actions'

export const initialState = {
  error: null,
  items: []
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case fetchSuccessItems:
      return {
        ...state,
        items: action.data
      }
    case fetchErrorItems:
      return {
        ...state,
        error: true
      }
    default:
      return {
        ...state
      }
  }
}
