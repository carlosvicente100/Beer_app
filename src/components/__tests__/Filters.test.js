import React from 'react'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import { render as rtlRender, fireEvent } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import { initialState as initialStateFilters } from '../../redux/reducers/filters'
import { initialState as initialStateItems } from '../../redux/reducers/items'
import reducer from '../../redux/reducers'

import Filters from '../Filters'

beforeEach(() => {
  jest.clearAllMocks()
})

const dummyProps = {
  updateFilters: jest.fn()
}
const initialReducerState = {
  filtersReducer: initialStateFilters,
  itemsReducer: initialStateItems
}

function render(ui, { initialState, store = createStore(reducer, initialState), ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

describe('Filters Component Suite', () => {
  test('Snapshot check', () => {
    expect(render(<Filters {...dummyProps} />, { initialReducerState }).baseElement).toMatchSnapshot()
  })

  describe('Filters Structure', () => {
    test('Should have 4 inputs with their titles and two buttons ', () => {
      const { container, getByText } = render(<Filters {...dummyProps} />)

      expect(getByText('Start Month')).toBeInTheDocument()
      expect(getByText('End Month')).toBeInTheDocument()
      expect(getByText('Start year')).toBeInTheDocument()
      expect(getByText('End year')).toBeInTheDocument()

      expect(container.getElementsByClassName('Select').length).toBe(4)

      expect(getByText('Filter')).toBeInTheDocument()
      expect(getByText('Filter')).toHaveAttribute('disabled')
      expect(getByText('Reset All')).toBeInTheDocument()
      expect(getByText('Filter')).toHaveAttribute('disabled')
    })
  })

  describe('Filters Functionallity', () => {
    test('Should reset Filters on Reset Button click', () => {
      const store = createStore(() => ({
        filtersReducer: {
          startDate: {
            year: 2020,
            month: 5
          },
          endDate: {
            year: 2040,
            month: 12
          }
        },
        itemsReducer: initialStateItems
      }))
      const { container, getByText } = render(<Filters {...dummyProps} />, { store })
      expect(getByText('Filter')).not.toHaveAttribute('disabled')
      fireEvent.click(getByText('Reset All'))
      expect(getByText('Filter')).toHaveAttribute('disabled')
      expect(dummyProps.updateFilters).toHaveBeenCalledTimes(1)
    })

    test('Should set Filters on Filter Button click', () => {
      const store = createStore(() => ({
        filtersReducer: {
          startDate: {
            year: 2020,
            month: 5
          },
          endDate: {
            year: 2040,
            month: 12
          }
        },
        itemsReducer: initialStateItems
      }))
      const { container, getByText } = render(<Filters {...dummyProps} />, { store })
      expect(getByText('Filter')).not.toHaveAttribute('disabled')
      fireEvent.click(getByText('Filter'))
      expect(getByText('Filter')).not.toHaveAttribute('disabled')
      expect(dummyProps.updateFilters).toHaveBeenCalledTimes(1)
    })
  })
})
