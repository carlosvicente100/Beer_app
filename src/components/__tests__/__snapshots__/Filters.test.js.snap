// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters Component Suite Snapshot check 1`] = `
.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  width: 100%;
}

.c3 {
  margin: auto;
}

.c0 {
  background-color: #227093;
  color: white;
  text-align: center;
}

.c1 {
  width: 50%;
  display: inline-grid;
  grid-area: input;
  grid-template-areas: 'span' 'select';
}

.c1 span {
  grid-area: span;
}

.c1 div.Select {
  width: 100%;
  grid-area: select;
}

.c4 {
  background-color: #2980b9;
  border: none;
  color: white;
  padding: 12px 16px;
  font-size: 16px;
  cursor: pointer;
}

.c4:hover {
  background-color: tomato;
}

.c4:disabled {
  background-color: #7f8c8d;
}

<body>
  <div>
    <div
      class="c0 Filters"
    >
      <label
        class="c1"
      >
        Start Month
        <div
          class="c2 Select"
        >
          <select
            class="c3"
            data-testid="select"
          >
            <option
              value="-1"
            >
              Empty
            </option>
            <option
              value="1"
            >
              January
            </option>
            <option
              value="2"
            >
              February
            </option>
            <option
              value="3"
            >
              March
            </option>
            <option
              value="4"
            >
              April
            </option>
            <option
              value="5"
            >
              May
            </option>
            <option
              value="6"
            >
              June
            </option>
            <option
              value="7"
            >
              July
            </option>
            <option
              value="8"
            >
              August
            </option>
            <option
              value="9"
            >
              September
            </option>
            <option
              value="10"
            >
              October
            </option>
            <option
              value="11"
            >
              November
            </option>
            <option
              value="12"
            >
              December
            </option>
          </select>
        </div>
      </label>
      <label
        class="c1"
      >
        End Month
        <div
          class="c2 Select"
        >
          <select
            class="c3"
            data-testid="select"
          >
            <option
              value="-1"
            >
              Empty
            </option>
            <option
              value="1"
            >
              January
            </option>
            <option
              value="2"
            >
              February
            </option>
            <option
              value="3"
            >
              March
            </option>
            <option
              value="4"
            >
              April
            </option>
            <option
              value="5"
            >
              May
            </option>
            <option
              value="6"
            >
              June
            </option>
            <option
              value="7"
            >
              July
            </option>
            <option
              value="8"
            >
              August
            </option>
            <option
              value="9"
            >
              September
            </option>
            <option
              value="10"
            >
              October
            </option>
            <option
              value="11"
            >
              November
            </option>
            <option
              value="12"
            >
              December
            </option>
          </select>
        </div>
      </label>
      <label
        class="c1"
      >
        Start year
        <div
          class="c2 Select"
        >
          <select
            class="c3"
            data-testid="select"
          >
            <option
              value="-1"
            >
              Empty
            </option>
          </select>
        </div>
      </label>
      <label
        class="c1"
      >
        End year
        <div
          class="c2 Select"
        >
          <select
            class="c3"
            data-testid="select"
          >
            <option
              value="-1"
            >
              Empty
            </option>
          </select>
        </div>
      </label>
      <button
        class="c4"
        disabled=""
      >
        Filter
      </button>
      <button
        class="c4"
        disabled=""
      >
        Reset All
      </button>
    </div>
  </div>
</body>
`;
